### ISSUE TITLE
[Finder] GitIgnore creates incorrect regex when ignoring 'everything in subdirectory but..'

### ISSUE THREAD

**Issue by Jeroeny** [symfony/symfony#37424]
**Symfony version(s) affected**: >=4.3

**Description**  

When you want to ignore everything in a certain directory, _except_ some specific files,
these are the `.gitignore` rules that can be used:
```gitignore
/example/**
!/example/example.txt
!/example/packages/
!/example/packages/example.yaml
``` 

You might think `/example/packages/` will be unignored by this, but that is not the case according to the `.gitignore` standard.
With this setup, any files in `/example/packages/` besides `/example/packages/example.yaml` will be ignored. But the `Gitignore::toRegex` does not match this logic properly.

See also: https://git-scm.com/docs/gitignore and https://stackoverflow.com/a/61556103/3265437

**How to reproduce**  
https://github.com/Jeroeny/reproduce/tree/gitign

**Possible Solution**  
The regex building logic would have to be adjusted. I've looked into this but wasn't able to fix it yet.

**Additional context**  
See the [reproducer](https://github.com/Jeroeny/reproduce/tree/gitign).

**Comment by Jeroeny** [comment]
Added a possible fix to the reproducer: https://github.com/Jeroeny/reproduce/blob/gitign/Gitignore.php

**Comment by stof** [comment]
I think the fix is not right either, because the way gitignore works is not that all negative rules (un-ignoring files again) are matched after positive rules (ignoring files). AFAIK, the order of rules is relevant.

**Comment by Jeroeny** [comment]
> I think the fix is not right either, because the way gitignore works is not that all negative rules (un-ignoring files again) are matched after positive rules (ignoring files). AFAIK, the order of rules is relevant.

Indeed, the order is very important for the final result. I'm still testing the possible solution a bit, but haven't found a case yet where it's incorrect.

**Comment by Jeroeny** [comment]
I had to change the way the regex is built up, to take lines' order into consideration. I made a new class in the reproducer that correctly matches files when they are re-ignored (or re-unignored) on lines after a previous line that (un-)ignored it: https://github.com/Jeroeny/reproduce/blob/gitign/GitignoreNew.php

The regex per positive `.gitignore` line is a group followed by a negative lookbehind with all the negations that were on lines áfter that line. 

It seems to work so far.
